core/src/impl/Kokkos_Error.hpp

```diff
@@ -69,8 +69,7 @@ namespace Impl {
 
 #if defined(KOKKOS_ENABLE_CUDA) && defined(__CUDA_ARCH__)
 
-#if defined(__APPLE__) || defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
-// cuda_abort does not abort when building for macOS.
+#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
```

Comment:
This comes from #3286
@ndellingwood would you please check if that is still an issue with our new minimum requirements?
I would much rather drop that silly workaround if we are able to.

Reply:
@dalg24 @masterleinad  I tested the PR sha 33038b6b855175cbe7a92a5bb2cd756f035cd7ca as-is with cuda/11.1+gcc/8.3.0 and I am seeing the warning:
```
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_ViewCtor.hpp(304): warning: missing return statement at end of non-void function "Kokkos::Impl::with_properties_if_unset(const Kokkos::Impl::ViewCtorProp<P...> &, const Property &, const Properties &...) [with P=<Kokkos::Cuda, Kokkos::CudaSpace, std::__cxx11::string, Kokkos::Impl::WithoutInitializing_t>, Property=Kokkos::CudaSpace::execution_space, Properties=<>]"
...
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_ViewCtor.hpp(339): warning: missing return statement at end of non-void function "Kokkos::Impl::get_property<Tag,P...>(const Kokkos::Impl::ViewCtorProp<P...> &) [with Tag=Kokkos::Impl::LabelTag, P=<Kokkos::Cuda, Kokkos::CudaSpace, std::__cxx11::string, Kokkos::Impl::WithoutInitializing_t>]"
...
```

I re-compiled with removal of the guards flagged in the message (line 72-79), leaving just
```
#define KOKKOS_IMPL_ABORT_NORETURN [[noreturn]]
```
and I'm seeing multiple warnings like
```
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Error.hpp(230): warning: function declared with "noreturn" does return
```

Reply:
@ndellingwood The question was if removing the workaround for `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` still breaks "CUDA random number generator tests on Pre-Volta archs" (#3286) and stuff in KokkosKernels (https://github.com/kokkos/kokkos-kernels/issues/785).

Reply:
The only pre-Volta architecture I have access to is Kepler but that is no longer a supported arch right?

Reply:
@ndellingwood Kepler 3.5 and later should still be supported in CUDA 11.0.

Reply:
I tested the PR after making these changes to remove the `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` guards:

```diff
diff --git a/core/src/impl/Kokkos_Error.hpp b/core/src/impl/Kokkos_Error.hpp
index 869bfb1..0447844 100644
--- a/core/src/impl/Kokkos_Error.hpp
+++ b/core/src/impl/Kokkos_Error.hpp
@@ -69,14 +69,8 @@ namespace Impl {
 
 #if defined(KOKKOS_ENABLE_CUDA) && defined(__CUDA_ARCH__)
 
-#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
-// required to workaround failures in random number generator unit tests with
-// pre-volta architectures
-#define KOKKOS_IMPL_ABORT_NORETURN
-#else
 // cuda_abort aborts when building for other platforms than macOS
 #define KOKKOS_IMPL_ABORT_NORETURN [[noreturn]]
-#endif
 
 #elif defined(KOKKOS_COMPILER_NVHPC)
```

with cuda/11.1+gcc/8.3.0 on Kepler35 with this configuration:
```bash
module load sems-gcc/8.3.0 sems-cuda/11.1.0 sems-cmake/3.23.1

cmake -DCMAKE_CXX_COMPILER=$KOKKOS_PATH/bin/nvcc_wrapper -DKokkos_ENABLE_CUDA=ON -DKokkos_ARCH_KEPLER35=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DKokkos_ENABLE_CUDA_LAMBDA=ON -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_BUILD_TYPE=RELEASE -DKokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON -DKokkos_ENABLE_DEPRECATION_WARNINGS=OFF -DKokkos_ENABLE_DEPRECATED_CODE_4=OFF $KOKKOS_PATH
```

All algorithms tests (including random num generator) passed, saw repeats of this warning during compilation:

```
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Error.hpp(231): warning: function declared with "noreturn" does return
```

Reply:
I think the crucial changes to test would be to remove the checks in `core/src/Cuda/Kokkos_Cuda_abort.hpp`, in particular activating  `while(true);`.

Reply:
@ndellingwood Could you test with the commit I pushed?

Reply:
Tests also passed when removing the `while(true);` hack, i.e.

```diff
diff --git a/core/src/Cuda/Kokkos_Cuda_abort.hpp b/core/src/Cuda/Kokkos_Cuda_abort.hpp
index 642581b..94cba76 100644
--- a/core/src/Cuda/Kokkos_Cuda_abort.hpp
+++ b/core/src/Cuda/Kokkos_Cuda_abort.hpp
@@ -81,10 +81,6 @@ __device__ inline void cuda_abort(const char *const message) {
   // __assertfail is not marked as [[noreturn]], even though it does not return.
   //  Disable with KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK to workaround failures
   //  in random number generator unit tests with pre-volta architectures
-#if !defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
-  while (true)
-    ;
-#endif
 }
 
 }  // namespace Impl
```

I'll test one more incremental change, removing the other `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` guard from `core/src/Cuda/Kokkos_Cuda_abort.hpp`

Reply:
Removing the last `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` guarded region from `core/src/Cuda/Kokkos_Cuda_abort.hpp `, that is this diff (in addition to the changes previously listed)

```diff
-// required to workaround failures in random number generator unit tests with
-// pre-volta architectures
-#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
-__device__ inline void cuda_abort(const char *const message) {
-#else
 [[noreturn]] __device__ inline void cuda_abort(const char *const message) {
-#endif
   const char empty[] = "";
 
   __assertfail((const void *)message, (const void *)empty, (unsigned int)0,
                (const void *)empty, sizeof(char));
```

lead to test failures in the algorithms tests, in particular with the random num tests:
```
    Start 1: KokkosAlgorithms_UnitTest_RandomAndSort

1: Test command: /ascldap/users/ndellin/kokkos/testing/Pr5441/algorithms/unit_tests/KokkosAlgorithms_UnitTest_RandomAndSort
1: Test timeout computed to be: 10000000
1: [==========] Running 5 tests from 1 test suite.
1: [----------] Global test environment set-up.
1: [----------] 5 tests from cuda
1: [ RUN      ] cuda.Random_XorShift64
1: Test Seed:1665600133315178171
1: Test Scalar=int
1:  -- Testing randomness properties
1: /ascldap/users/ndellin/kokkos/algorithms/unit_tests/TestRandom.hpp:310: Failure
1: Expected: (std::abs(mean_eps)) < (tolerance), actual: 0.0406466 vs 0.000139188
1: /ascldap/users/ndellin/kokkos/algorithms/unit_tests/TestRandom.hpp:311: Failure
1: Expected: (std::abs(variance_eps)) < (1.5 * tolerance), actual: 0.0406786 vs 0.000208782
1:  -- Testing 1-D histogram
1: cudaStreamSynchronize(stream) error( cudaErrorIllegalInstruction): an illegal instruction was encountered /ascldap/users/ndellin/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:172
1: Backtrace:
1:                         [0x55f293]
1:                         [0x5577c8]
1:                         [0x55782b]
1:                         [0x5676fe]
1:                         [0x568271]
1:                         [0x4328ca]
1:                         [0x4344d3]
1:                         [0x46884c]
1:                         [0x46abf9]
1:                         [0x53a21a]
1:                         [0x529268]
1:                         [0x52968a]
1:                         [0x529c0d]
1:                         [0x531aea]
1:                         [0x531f78]
1:                         [0x410b50]
1: __libc_start_main [0x7f8e0e9bf555]
1:                         [0x4156ed]
1/1 Test #1: KokkosAlgorithms_UnitTest_RandomAndSort ...Subprocess aborted***Exception:   6.72 sec
```